### PR TITLE
CDAP-3605 Intercept entitiy deletion to also remove metadata

### PR DESCRIPTION
**Pull request by henry-cask**
Intercept the entity deletion of App, Program, Dataset, and Stream so it also remove the corresponding business metadata (property and tags).

Add Dataset service and transaction manager to some tests.

### CODE REVIEW

**Comment by henry-cask** [comment]
Build: http://builds.cask.co/browse/CDAP-DUT2851-5

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java [hunk 1]
```diff
@@ -695,6 +702,75 @@ void deleteApp(Id.Application appId, ApplicationSpecification spec) throws Excep
...
+    // TODO: Need to remove this we support prefix search of metadata type.
+    ApplicationSpecification appSpec = store.getApplication(appId);
+    for (ProgramType programType : ProgramType.values()) {
+      deleteProgramBusinessMetadataForApp(programType, appId, appSpec);
+    }
+  }
+
+  /**
+   *  Delete the business metadata for a program type in an application.
+   *  TODO: Need to remove this we support prefix search of metadata type.
```
Comment (sreevatsanraman) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java]: Open a JIRA for 3.2 and add that to the TODO.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java [hunk 2]
```diff
@@ -695,6 +702,75 @@ void deleteApp(Id.Application appId, ApplicationSpecification spec) throws Excep
...
+  /**
+   * Delete the business metadata for the application and the programs.
+   */
+  private void deleteAppBusinessMetadata(Id.Application appId) {
+    // Remove metadata for the Application itself.
+    businessMds.removeMetadata(appId);
+
+    // Remove business metadata for the programs of the Application
+    // TODO: Need to remove this we support prefix search of metadata type.
+    ApplicationSpecification appSpec = store.getApplication(appId);
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java]: the app spec is available in the caller, can just pass it in.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java [hunk 3]
```diff
@@ -686,6 +691,8 @@ void deleteApp(Id.Application appId, ApplicationSpecification spec) throws Excep
     }
     deleteProgramLocations(appId);
 
+    deleteAppBusinessMetadata(appId);
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java]: this can get called during cdap upgrade. Will everything work in the upgrade tool?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java [hunk 4]
```diff
@@ -695,6 +702,75 @@ void deleteApp(Id.Application appId, ApplicationSpecification spec) throws Excep
...
+  }
+
+  /**
+   *  Delete the business metadata for a program type in an application.
+   *  TODO: Need to remove this we support prefix search of metadata type.
+   */
+  private void deleteProgramBusinessMetadataForApp(ProgramType programType, Id.Application appId,
+                                                   ApplicationSpecification appSpec) {
+    Set<String> programNames = null;
+    switch (programType) {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java]: seems a bit odd to loop through program types and then switch on them.  For each app delete call, there will be guaranteed 2 log info messages about webapp and custom action.  Simpler if we get all flows, add to a `Set<Id.Program>`, get all mapreduce, add to set, etc.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java [hunk 5]
```diff
@@ -125,12 +128,21 @@ public boolean hasType(Id.DatasetType datasetTypeId) throws DatasetManagementExc
   @Override
   public void deleteInstance(Id.DatasetInstance datasetInstanceId)
     throws DatasetManagementException, IOException, ServiceUnavailableException {
+    // Remove metadata for the dataset
+    businessMds.removeMetadata(datasetInstanceId);
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java]: it would be better if these operations happened in the same transaction.  Guessing it's too late now, but do we have a jira open for it?

**Comment by albertshau** [comment]
couple small comments and one question about upgrade

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java (hunk 2 above)
Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java]: Will do

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java (hunk 4 above)
Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java]: Sounds good to me. Will make the change

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java (hunk 3 above)
Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java]: I am not sure what impact that upgrade tool has on this. Is this covered by any test?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java (hunk 5 above)
Comment (henry-cask) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java]: No JIRA yet, will file one for improvement.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java (hunk 1 above)
Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java]: https://issues.cask.co/browse/CDAP-3669

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java (hunk 3 above)
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java]: there's an upgrade test in ApplicationLifecycleServiceTest, but I'm guessing it already sets up the business mds.  I think there are some integration tests but those aren't in the cdap repo.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java (hunk 5 above)
Comment (henry-cask) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java]: JIRA filed:
https://issues.cask.co/browse/CDAP-3670

**Comment by henry-cask** [comment]
Build:
http://builds.cask.co/browse/CDAP-DUT2851-7

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java [hunk 6]
```diff
@@ -232,9 +239,30 @@ public void testLineage() throws Exception {
     Assert.assertEquals(400, fetchLineage(datasetInstance, 100, 200, -10).getResponseCode());
   }
 
-  @After
-  public void after() throws Exception {
-    deleteApp(application, 200);
```
Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: why is this removed?

File: cdap-data-fabric-tests/src/test/java/co/cask/cdap/data/stream/InMemoryStreamCoordinatorClientTest.java [hunk 7]
```diff
@@ -47,7 +46,6 @@
 
   @BeforeClass
   public static void init() throws IOException {
-    cConf.set(Constants.CFG_LOCAL_DATA_DIR, tmpFolder.newFolder().getAbsolutePath());
```
Comment (bdmogal) [inline cdap-data-fabric-tests/src/test/java/co/cask/cdap/data/stream/InMemoryStreamCoordinatorClientTest.java]: why is this removed?

File: cdap-data-fabric-tests/src/test/java/co/cask/cdap/data/stream/LocalStreamFileJanitorTest.java [hunk 8]
```diff
@@ -57,7 +56,6 @@
 
   @BeforeClass
   public static void init() throws IOException {
-    cConf.set(Constants.CFG_LOCAL_DATA_DIR, tmpFolder.newFolder().getAbsolutePath());
```
Comment (bdmogal) [inline cdap-data-fabric-tests/src/test/java/co/cask/cdap/data/stream/LocalStreamFileJanitorTest.java]: ditto

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/DistributedBusinessMetadataStore.java [hunk 9]
```diff
@@ -0,0 +1,460 @@
...
+    Id.DatasetInstance.from(Id.Namespace.SYSTEM, "business.metadata");
+  private static final Map<String, String> EMPTY_PROPERTIES = ImmutableMap.of();
+  private static final Set<String> EMPTY_TAGS = ImmutableSet.of();
+
+  private final Transactional<BusinessMdsIterable, BusinessMetadataDataset> txnl;
+  private final CConfiguration cConf;
+  private final MetadataChangePublisher changePublisher;
+
+  @Inject
+  DistributedBusinessMetadataStore(TransactionExecutorFactory txExecutorFactory,
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/DistributedBusinessMetadataStore.java]: this is used in standalone as well. should we call it ``DefaultBusinessMetadataStore``

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/DistributedBusinessMetadataStore.java [hunk 10]
```diff
@@ -0,0 +1,460 @@
...
+        input.businessMds.removeProperties(entityId);
+        input.businessMds.removeTags(entityId);
+        return null;
+      }
+    });
+    MetadataRecord previous = previousRef.get();
+    publish(previous, new MetadataRecord(entityId), new MetadataRecord(previous));
+  }
+
+  void removeMetadataNoPublish(final Id.NamespacedId entityId) {
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/DistributedBusinessMetadataStore.java]: private

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/DistributedBusinessMetadataStore.java [hunk 11]
```diff
@@ -0,0 +1,460 @@
...
+                                           input.businessMds.getTags(entityId)));
+        input.businessMds.removeProperties(entityId);
+        return null;
+      }
+    });
+    publish(previousRef.get(), new MetadataRecord(entityId),
+            new MetadataRecord(entityId, previousRef.get().getProperties(), EMPTY_TAGS));
+  }
+
+  void removePropertiesNoPublish(final Id.NamespacedId entityId) {
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/DistributedBusinessMetadataStore.java]: private

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/DistributedBusinessMetadataStore.java [hunk 12]
```diff
@@ -0,0 +1,460 @@
...
+                                           input.businessMds.getTags(entityId)));
+        input.businessMds.removeTags(entityId);
+        return null;
+      }
+    });
+    MetadataRecord previous = previousRef.get();
+    publish(previous, new MetadataRecord(entityId), new MetadataRecord(entityId, EMPTY_PROPERTIES, previous.getTags()));
+  }
+
+  void removeTagsNoPublish(final Id.NamespacedId entityId) {
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/DistributedBusinessMetadataStore.java]: private

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/InMemoryBusinessMetadataStore.java [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+import co.cask.cdap.proto.metadata.MetadataRecord;
+import co.cask.cdap.proto.metadata.MetadataSearchTargetType;
+
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Implementation of {@link BusinessMetadataStore} used in memory mode.
+ */
+public class InMemoryBusinessMetadataStore implements BusinessMetadataStore {
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/InMemoryBusinessMetadataStore.java]: Should we call this ``NoOpBusinessMetadataStore``?

**Comment by bdmogal** [comment]
@henry-cask some minor comments. None of my comments should require a full build if you already have one green. However, I highly recommend getting the guice stuff confirmed with @poornachandra

**Comment by sreevatsanraman** [comment]
Closing this PR - re-open against release/3.2

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java (hunk 6 above)
Comment (henry-cask) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: No it was move up right just after the Before. Sorry about the confusion.

File: cdap-data-fabric-tests/src/test/java/co/cask/cdap/data/stream/LocalStreamFileJanitorTest.java (hunk 8 above)
Comment (henry-cask) [inline cdap-data-fabric-tests/src/test/java/co/cask/cdap/data/stream/LocalStreamFileJanitorTest.java]: Gah accidental delete.

File: cdap-data-fabric-tests/src/test/java/co/cask/cdap/data/stream/InMemoryStreamCoordinatorClientTest.java (hunk 7 above)
Comment (henry-cask) [inline cdap-data-fabric-tests/src/test/java/co/cask/cdap/data/stream/InMemoryStreamCoordinatorClientTest.java]: Accidental removed.
